utils/merge_syft_sbom.py

```diff
@@ -0,0 +1,128 @@
+    list of non-Pypi components from Cachi2's SBOM, then remove the ones reported
+    by Syft in the merged SBOM.
+    """
+    return (
+        component.get("purl", "").startswith("pkg:pypi")
+        and "version" not in component
+    )
+
+
+# Unique key functions
+def _unique_key_cachi2(component: dict) -> str:
+    """Create a unique key from Cachi2 reported components."""
+    url = urlsplit(component["purl"])
+    # omit the query part, since Cachi2 sets it and Syft does not
+    return url.scheme + ":" + url.path
+
+
+def _unique_key_syft(component: dict) -> str:
+    """Create a unique key for Syft reported components."""
+    # very few Syft components don't have a purl (e.g. type=OS)
+    if "purl" not in component:
+        return component.get('name', '') + '@' + component.get('version', '')
+
+    if "@" in component["purl"]:
+        parts = component["purl"].split("@")
+        # We need to encode special characters in the version part,
+        # and also need to ensure the name is all lowercase
+        return parts[0].lower() + "@" + quote_plus(parts[1])
+    else:
+        return component["purl"]
+
+
+def _get_syft_component_filter(cachi_sbom_components: list[dict[str, Any]]) -> Callable:
+    """
+    Get a filter function for Syft components in the SBOM.
+
+    This function applies several rules to determine if a component should be
+    removed from the final merged SBOM.
+    """
+    cachi2_non_pypi_components = [
+        component["name"] for component in cachi_sbom_components
+        if _is_cachi2_non_pypi_pip_dependency(component)
+    ]
+
+    cachi2_indexed_components = {
+        _unique_key_cachi2(component): component
+        for component in cachi_sbom_components
+    }
+
+    def is_duplicate_non_pypi_component(component: dict[str, Any]) -> bool:
+        return component["name"] in cachi2_non_pypi_components
+
+    def component_is_duplicated(component: dict[str, Any]) -> bool:
+        key = _unique_key_syft(component)
+
+        return _is_syft_local_golang_component(component) or\
+            is_duplicate_non_pypi_component(component) or\
+            key in cachi2_indexed_components.keys()
+
+    return component_is_duplicated
+
+
+def merge_sboms(cachi2_sbom_path: str, syft_sbom_path: str) -> str:
+    """Merge Cachi2 components into the Syft SBOM while removing duplicates."""
+    with open(cachi2_sbom_path) as file:
+        cachi2_sbom = json.load(file)
+
+    with open(syft_sbom_path) as file:
+        syft_sbom = json.load(file)
+
+    is_duplicate_component = _get_syft_component_filter(cachi2_sbom["components"])
+
+    filtered_syft_components = [
+        component for component in syft_sbom["components"]
+        if not is_duplicate_component(component)
+    ]
+
+    syft_sbom["components"] = filtered_syft_components + cachi2_sbom["components"]
+    syft_sbom["metadata"]["tools"].append({
+        "vendor": "Red Hat",
+        "name": "Cachi2"
+    })
```

Comment:
IMO this doesn't belong in a merge script but in cachi2 itself

Reply:
Agreed. We will need to do some merging magic here regardless.

Reply:
I changed the script so it will merge any `metadata.tools` if they're available. The second part should be solved by #232.